### PR TITLE
Revert "dont add special case for warning"

### PR DESCRIPTION
**Pull request by CrowdHailer**
This reverts commit e5eb3f254ef3b2ca394920f654900c1040ef53a0.

This was the changes I look out of pr #19 
I think solving this issue #18 is a good goal but I would like to take the time for a more complete solution and this PR is to keep those efforts in one place

### CODE REVIEW

**Comment by CrowdHailer** [comment]
@bill-mybiz I read the long comment you put on the last PR again.

you are right that there are only three return values to consider. however because the macros are execute before the functions they cannot inspect the return types of those functions.

```elixir
def ok_func do
  {:ok, :hi}
end

OK.with do
  # stuff
  ok_func()
end
```

I believe this would still throw a warning, but when the macro runs it can't know that `ok_func` always returns an ok tuple
```

```

**Comment by bill-mybiz** [comment]
> I believe this would still throw a warning, but when the macro runs it can't know that ok_func always returns an ok tuple

```
  test "ok_func test" do
    OK.with do
      :hi <- ok_func()
      ok_func()
    end
  end

  def ok_func() do
    {:ok, :hi}
  end
```

Does _not_ produce a compiler warning (I also checked this with `defmacro ok_func`). This is because it does not match against the added `nest/1` clauses, rather it matches against `defp nest([normal | []]) do`. 

You can check this yourself if you would like, all you need to do is pull the code down and run `mix test`. There is only one generated warning, and it is appropriate for that warning because `{:x, a}` is returned (and will never match).

That said, I totally get wanting to have a different solution if you don't feel right with the approach, since obviously it's your :baby_bottle: :baby: ! :laughing: 

Like I said before, I love the work that you continue to put into this lib. And even more so now that I'm using it more and more in my code. It's freaking awesome!! Like I've said on EF, I use `with` _a lot_, and my code is _ridiculously_ more readable. [Here](https://github.com/ibgib/ibgib/blob/0039201d5f731026a060629a1317d7dcf7a074b8/ib_gib_umb/apps/web_gib/lib/web_gib/oy.ex#L42) is just one of many times I've already used it. And I'll point out one other thing in that particular code block: I've also combined it with `~>>` and it looks to be working really nicely. I get the convenience of the pipe operator with the structure of the ok/error result tuples. And that looks so much cleaner. :+1: :eyes:

**Comment by CrowdHailer** [comment]
I really like this line.
```elixir
else
      reason -> OK.failure handle_ok_error(reason, log: true)
end
```

Looks very similar to something I am trying to achieve with my web framework.

**Comment by CrowdHailer** [comment]
OK, turns out you are write about the function point. Shows me for being lazy and not spinning up a box to test it on v 1.4. I'm sure I had managed to get an error for functions that only return one value. But if i can't reproduce it, it doesn't make sense to take it into consideration

File: test/ok_test.exs [hunk 1]
```diff
@@ -131,7 +131,7 @@ defmodule OKTest do
     assert_raise CaseClauseError, fn() ->
       OK.with do
         a <- safe_div(8, 2)
-        {:x, a}
+        (fn() -> {:x, a} end).()
```
Comment (CrowdHailer) [inline test/ok_test.exs]: Nice trick based off the functions to remove the last warning I had in tests

Comment (bill-mybiz) [inline test/ok_test.exs]: :+1:

**Comment by bill-mybiz** [comment]
> I really like this line.
> ```
> else
>   reason -> OK.failure handle_ok_error(reason, log: true)
> end
> ```

It's done with a macro so that when an error occurs, it logs the error within the function that has this - at least I thought that was the case. With 1.5.1, it is now giving me a different warning! :sweat_smile: The warning could have been there before and I didn't notice it (I'm still pretty swamped with 1.4 warnings all over...)

It's taking a little while for me to understand your exception handling in the `with` macro. I know that it allows for passing through of errors, but I'm just not understanding all aspects just yet. 

The following repros the issue:
```
  test "checking default handler" do
    OK.with do
      a <- safe_div(8, 2)
      _ <- safe_div(a, 0)
    else
      reason -> # warning on this line
        :some_other_reason
    end
  end
```

If anything this could be a separate issue. As far as the PR goes, it LGTM :+1: Those warnings are gone for me in both my project and when I updated my fork of your lib. :tada:

**Comment by CrowdHailer** [comment]
I think this warning is because your else block catches all possibilities. in essence the macro adds a final catch all clause to pass through any errors not dealt with. if you add any guard clauses to the line you commented I think the warning should go away

**Comment by bill-mybiz** [comment]
> I think this warning is because your else block catches all possibilities. in essence the macro adds a final catch all clause to pass through any errors not dealt with.

I think this is correct.

> if you add any guard clauses to the line you commented I think the warning should go away

I tested this and you're correct, but I don't really want to go adding unnecessary guards! :smile:

I tried figuring out how to only add the catchall (in `with/2`) when `exceptional` doesn't already have a catchall, but I'm not quite sure how to approach it. If you know how to do this off the top of your head (maybe a pattern match?), then I would recommend that for a fix, which begs the question of creating an issue for this particular warning. I think the expected behavior should be to allow a catchall as I'm doing. 

Btw, I did retest the actual functionality, and it works just fine - which makes sense as you are adding the catchall _after_ the given `exceptional` block.  It's just the warning on the unused catchall.

**Comment by CrowdHailer** [comment]
Don't know how to do this off the top of my head, sure can work it out eventually. It's certainly a new issue, although not one that concerns me too much right now.

**Pull request merged**